# Day 5: Hydrothermal Venture

## Part one

You come across a field of hydrothermal vents on the ocean floor! These vents constantly produce large, opaque clouds, so it would be best to avoid them if possible.

They tend to form in lines; the submarine helpfully produces a list of nearby lines of vents (your puzzle input) for you to review. For example:

```
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
```

Each line of vents is given as a line segment in the format `x1,y1 -> x2,y2` where `x1,y1` are the coordinates of one end the line segment and `x2,y2` are the coordinates of the other end. These line segments include the points at both ends. In other words:

- An entry like `1,1 -> 1,3` covers points `1,1, 1,2, and 1,3`.
- An entry like `9,7 -> 7,7` covers points `9,7, 8,7, and 7,7`.

For now, only consider horizontal and vertical lines: lines where either `x1 = x2` or `y1 = y2`.

So, the horizontal and vertical lines from the above list would produce the following diagram:

```
.......1..
..1....1..
..1....1..
.......1..
.112111211
..........
..........
..........
..........
222111....
```

In this diagram, the top left corner is 0,0 and the bottom right corner is 9,9. Each position is shown as **the number of lines which cover that point** or `.` if no line covers that point. The top-left pair of `1s`, for example, comes from `2,2 -> 2,1`; the very bottom row is formed by the overlapping lines `0,9 -> 5,9` and `0,9 -> 2,9`.

To avoid the most dangerous areas, you need to determine **the number of points where at least two lines overlap**. In the above example, this is anywhere in the diagram with a 2 or larger - a total of 5 points.

Consider only horizontal and vertical lines. **At how many points do at least two lines overlap?**

In [1]:
class Map:
    def __init__(self, size):
        self.size = size
        self.board = [[]]
        self.init_board()
        
    def init_board(self):
        board = []
        for x in range(0, self.size):
            line = []
            for y in range(0, self.size):
                line.append(0)
            board.append(line)
        self.board = board
        
    def check_board(self, x, y):
        self.board[x][y] += 1
    
    def print_board(self):
        print('\n'.join(' '.join(map(str, x)) for x in self.board))
        
    def count_overlap(self):
        count = 0
        for line in self.board:
            for val in line:
                if val > 1:
                    count += 1
        
        return count
                
    
demo_board = Map(10)
demo_board.check_board(0, 1)
demo_board.check_board(0, 1)
demo_board.check_board(9, 9)
demo_board.check_board(9, 9)
demo_board.print_board()
demo_board.count_overlap()

0 2 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 2


2

In [2]:
class LineOfVent:
    
    def __init__(self, x1, y1, x2, y2):
        if x1 < x2:
            self.x1 = x1
            self.x2 = x2
        else:
            self.x1 = x2
            self.x2 = x1
        
        if y1 < y2:
            self.y1 = y1
            self.y2 = y2
        else:
            self.y1 = y2
            self.y2 = y1

    
    def __str__(self):
        return '(x1: %s - y1: %s) - (x2: %s - y2: %s)' % (self.x1, self.y1, self.x2, self.y2, )

    
    
    def is_vertical_or_horizontal(self):
        return (self.x1 == self.x2) or (self.y1 == self.y2)
    
    
    def generate_fields(self):
        field = []
        for x in range(self.x1, self.x2 + 1):
            for y in range(self.y1, self.y2 + 1):
                field.append([x, y])
        return field
    
line_of_vent_1 = LineOfVent(1, 2, 3, 4)
print('%s' % (str(line_of_vent_1), ))
if line_of_vent_1.is_vertical_or_horizontal(): print('line of vent is hor or ver')
line_of_vent_1.generate_fields()

(x1: 1 - y1: 2) - (x2: 3 - y2: 4)


[[1, 2], [1, 3], [1, 4], [2, 2], [2, 3], [2, 4], [3, 2], [3, 3], [3, 4]]

In [3]:
import re

line_of_vents = []

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        m = re.search('^([0-9]+),([0-9]+) -> ([0-9]+),([0-9]+)$', line)
        line_of_vents.append(LineOfVent(int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))))
        
for i in range (0, 5):
    print('%s - hor/ver: %s' % (str(line_of_vents[i]), line_of_vents[i].is_vertical_or_horizontal(), ))

(x1: 194 - y1: 556) - (x2: 739 - y2: 556) - hor/ver: True
(x1: 818 - y1: 524) - (x2: 818 - y2: 920) - hor/ver: True
(x1: 340 - y1: 300) - (x2: 774 - y2: 734) - hor/ver: False
(x1: 146 - y1: 434) - (x2: 223 - y2: 511) - hor/ver: False
(x1: 122 - y1: 47) - (x2: 841 - y2: 766) - hor/ver: False


In [4]:
ex1_board = Map(1000)

for line_of_vent in line_of_vents:
    if line_of_vent.is_vertical_or_horizontal():
        for field in line_of_vent.generate_fields():
            ex1_board.check_board(field[0], field[1])
            
ex1_board.count_overlap()

7085

Your puzzle answer was `7085`.

## Part two

Unfortunately, considering only horizontal and vertical lines doesn't give you the full picture; you need to also consider **diagonal lines**.

Because of the limits of the hydrothermal vent mapping system, the lines in your list will only ever be horizontal, vertical, or a diagonal line at exactly 45 degrees. In other words:

- An entry like `1,1 -> 3,3` covers points `1,1`, `2,2`, and `3,3`.
- An entry like `9,7` -> `7,9` covers points `9,7`, `8,8`, and `7,9`.

Considering all lines from the above example would now produce the following diagram:

```
1.1....11.
.111...2..
..2.1.111.
...1.2.2..
.112313211
...1.2....
..1...1...
.1.....1..
1.......1.
222111....
```

You still need to determine **the number of points where at least two lines overlap**. In the above example, this is still anywhere in the diagram with a 2 or larger - now a total of `12` points.

Consider all of the lines. **At how many points do at least two lines overlap?**

In [12]:
import operator

class LineOfVentEx2:
    
    def __init__(self, x1, y1, x2, y2):
        self.x1 = x1
        self.x2 = x2
        
        self.y1 = y1
        self.y2 = y2

        
    def __str__(self):
        return '(x1: %s - y1: %s) - (x2: %s - y2: %s)' % (self.x1, self.y1, self.x2, self.y2, )

        
    def is_horizontal(self):
        return (self.x1 == self.x2)

        
    def is_vertical(self):
        return (self.y1 == self.y2)
    
    
    def is_diagonal(self):
        return abs(self.x1 - self.x2) == abs(self.y1 - self.y2)
    
    
    def is_horizontal_vertical_diagonal(self):
        return self.is_horizontal() or self.is_vertical() or self.is_diagonal()
    
    
    def generate_line(self):
        line = []
        
        if self.is_horizontal():
            line.append([self.x1, self.y1])
            if self.y1 < self.y2:
                for y in range(self.y1 + 1, self.y2 + 1):
                    line.append([self.x1, y])
            else:
                for y in range(self.y2, self.y1):
                    line.append([self.x1, y])
        
        elif self.is_vertical():
            line.append([self.x1, self.y1])
            if self.x1 < self.x2:
                for x in range(self.x1 + 1, self.x2 + 1):
                    line.append([x, self.y1])
            else:
                for x in range(self.x2, self.x1):
                    line.append([x, self.y1])
        
        elif self.is_diagonal():
            line.append([self.x1, self.y1])
            if self.x1 < self.x2:
                x_func = operator.add
            else:
                x_func = operator.sub
                
            if self.y1 < self.y2:
                y_func = operator.add
            else:
                y_func = operator.sub
                
            x = x_func(self.x1, 1)
            y = y_func(self.y1, 1)
            line.append([x, y])
            
            while x != self.x2:
                x = x_func(x, 1)
                y = y_func(y, 1)
                line.append([x, y])
        
        return line
    
line_of_vent_2 = LineOfVentEx2(1, 3, 4, 6)
print('%s' % (str(line_of_vent_2), ))
if line_of_vent_2.is_vertical(): print('line of vent is vertical')
if line_of_vent_2.is_horizontal(): print('line of vent is horizontal')
if line_of_vent_2.is_diagonal(): print('line of vent is diagonal')
line_of_vent_2.generate_line()

(x1: 1 - y1: 3) - (x2: 4 - y2: 6)
line of vent is diagonal


[[1, 3], [2, 4], [3, 5], [4, 6]]

In [18]:
import re

line_of_vents_ex2 = []

with open('data/input.txt', 'r') as f:
    for line in f.read().splitlines():
        m = re.search('^([0-9]+),([0-9]+) -> ([0-9]+),([0-9]+)$', line)
        line_of_vents_ex2.append(LineOfVentEx2(int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))))
        
for i in range (0, 5):
    print('%s - hor/ver/diag: %s' % (str(line_of_vents_ex2[i]), line_of_vents_ex2[i].is_horizontal_vertical_diagonal(), ))

(x1: 194 - y1: 556) - (x2: 739 - y2: 556) - hor/ver/diag: True
(x1: 818 - y1: 920) - (x2: 818 - y2: 524) - hor/ver/diag: True
(x1: 340 - y1: 734) - (x2: 774 - y2: 300) - hor/ver/diag: True
(x1: 223 - y1: 511) - (x2: 146 - y2: 434) - hor/ver/diag: True
(x1: 841 - y1: 47) - (x2: 122 - y2: 766) - hor/ver/diag: True


In [19]:
ex2_board = Map(1000)

for line_of_vent_ex2 in line_of_vents_ex2:
    if line_of_vent_ex2.is_horizontal_vertical_diagonal():
        for field in line_of_vent_ex2.generate_line():
            ex2_board.check_board(field[0], field[1])
            
ex2_board.count_overlap()

20271

Your puzzle answer was 20271.